zerver/models.py

```diff
@@ -183,34 +180,89 @@ def stream_cache_transformer(stream: Stream) -> str:
         cache_transformer=stream_cache_transformer,
     )  # type: Dict[int, str]
 
-    def personal_query_function(recipient_ids: List[int]) -> List[UserProfile]:
-        user_ids = [
-            recipient_id_to_type_pair_dict[recipient_id][1]
-            for recipient_id in recipient_ids
-        ]
-        return UserProfile.objects.filter(id__in=user_ids)
+    # Now we have to create display_recipients for personal and huddle messages.
+    # We do this via generic_bulk_cached_fetch, supplying apprioprate functions to it.
 
-    def personal_id_fetcher(user: UserProfile) -> int:
-        return type_pair_to_recipient_id_dict[(Recipient.PERSONAL, user.id)]
+    def personal_and_huddle_query_function(recipient_ids: List[int]) -> List[Tuple[int, List[UserProfile]]]:
+        """
+        Return a list of tuples of the form (recipient_id, [list of UserProfiles])
+        where [list of UserProfiles] has users corresponding to the recipient,
+        so the receiving userin Recipient.PERSONAL case,
+        or in Personal.HUDDLE case - users in the huddle.
+        This is a pretty hacky return value, but it needs to be in this form,
+        for this function to work as the query_function in generic_bulk_cached_fetch.
+        """
 
-    def personal_cache_transformer(user_profile: UserProfile) -> DisplayRecipientCacheT:
-        return [user_profile_to_display_recipient_dict(user_profile)]
+        def bulk_get_user_profile_by_id(uids: List[int]) -> Dict[int, UserProfile]:
+            return generic_bulk_cached_fetch(
+                # Use a separate cache key to protect us from conflicts with
+                # the get_user_profile_by_id cache.
+                # (Since we fetch without select_related() here)
+                lambda uid: 'bulk_fetch_display_recipients:' + user_profile_by_id_cache_key(uid),
```

Comment:
We should make this a proper cache function in `cache.py`, since that serves in part as our registry of cache keys used.

Reply:
Done in the last commit, extracting the function.